### PR TITLE
1660 - Add reduced data filter while getting data in merge job within Estimates by Job Role pipeline

### PR DESCRIPTION
**Pull request by AchyutRatkanthiwar**
## Description
Trello ticket [#1660](https://trello.com/c/1LzmUu8U/1660-add-filter-for-reduced-dataset-for-job-role-pipeline)

Added a reduced data filter while getting data in merge job within Estimates by Job Role pipeline. The data is filtered based in `cqc_location_import_date` column. monthly data for last 2 years is taken and only quarterly data before that. 

With the reduced data, validation of merged metadata was failing as the data now did not have records with `ascwds_filtering_rule` set to 'contained_invalid_missing_data_code'. So this is now handled by subtracting it while validation of merged metadata. the dataclass is not updated directly as this is still needed in IND CQC pipeline.

## Testing
- [x] Unit tests passing
- [x] Successful [Step Function run](https://eu-west-2.console.aws.amazon.com/states/home?region=eu-west-2#/v2/executions/details/arn:aws:states:eu-west-2:856699698263:execution:1660-jr-reduced-data-Ind-CQC-Filled-Post-Estimates-By-Role:f2e1c11e-46b5-4c31-bd8e-2c562a87ba2e)
- [x] Outputs checked in Athena

Outputs to compare branch and main are attached to the trello ticket.

## Checklist (delete if not relevant)
- [x] Unit tests added/amended
- [x] Docstrings added/updated
- [x] Updated CHANGELOG
- [x] Code reviewed by AI
- [x] Moved Trello ticket to PR column

## Reviewer checklist
- [ ] PR solves the Trello ticket requirements
- [ ] Outputs appear reasonable and understandable
- [ ] The overall approach is appropriate and not over-engineered
- [ ] No obvious scalability, performance or interdependency concerns
- [ ] Tests appropriately cover the main behaviour/edge cases
- [ ] Naming and structure are broadly understandable and consistent
- [ ] Docstrings are sufficient for future maintenance
- [ ] Docstrings cover non-obvious behaviour

### CODE REVIEW

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/utils/merge_utils.py [hunk 1]
```diff
@@ -55,3 +57,60 @@ def join_estimates_to_ascwds(
...
+    Returns:
+        pl.Expr: A Polars boolean expression that can be used inside `.filter()` or
+            `.with_columns()` to select rows based on the reduced data strategy.
+    """
+    today = today or date.today()
+
+    fy_year = today.year if today.month >= fy_start_month else today.year - 1
+    current_fy_start = date(fy_year, fy_start_month, 1)
+
+    monthly_start = date(current_fy_start.year - lookback_fy_years, fy_start_month, 1)
```
Comment (RoyPriceCS50) [inline projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/utils/merge_utils.py]: Do you need current_fy_start ?

You could do:
monthly_start = date(fy_year.year - lookback_fy_years, fy_start_month, 1)

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/utils/merge_utils.py [hunk 2]
```diff
@@ -55,3 +57,60 @@ def join_estimates_to_ascwds(
...
+        fy_start_month (int): Month in which the financial year starts
+            (default is 4 for April).
+
+        lookback_fy_years (int): Number of financial years to retain in full before
+            applying sampling.
+
+        quarter_months (tuple[int, ...]): Months considered valid for quarterly sampling
+            of historical data.
+
+        col (str): Name of the date column the filter is applied to.
```
Comment (RoyPriceCS50) [inline projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/utils/merge_utils.py]: Make this name more explicit.

date_col

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/utils/merge_utils.py [hunk 3]
```diff
@@ -55,3 +57,60 @@ def join_estimates_to_ascwds(
...
+        today (date | None): Reference date used to compute financial year boundaries.
+            If None, defaults to the current system date.
+
+        fy_start_month (int): Month in which the financial year starts
+            (default is 4 for April).
+
+        lookback_fy_years (int): Number of financial years to retain in full before
+            applying sampling.
+
+        quarter_months (tuple[int, ...]): Months considered valid for quarterly sampling
```
Comment (MarionJHolloway) [inline projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/utils/merge_utils.py:96]: "Defaults to Jan, Apr, Jul, and Oct."

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/utils/merge_utils.py [hunk 4]
```diff
@@ -55,3 +57,60 @@ def join_estimates_to_ascwds(
...
+            of historical data.
+
+        col (str): Name of the date column the filter is applied to.
+
+    Returns:
+        pl.Expr: A Polars boolean expression that can be used inside `.filter()` or
+            `.with_columns()` to select rows based on the reduced data strategy.
+    """
+    today = today or date.today()
```
Comment (MarionJHolloway) [inline projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/utils/merge_utils.py]: Might be worth typing this to be clear it's not a bool

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_01_merge_metadata.py [hunk 5]
```diff
@@ -116,14 +120,20 @@ def main(
...
-    )
+    ).filter(reduced_data_filter_expr())
 
     source_df = add_list_column_validation_check_flags(
         source_df, [IndCqcColumns.services_offered]
     )
 
     print(f"source df schema: {source_df.schema}")
 
+    allowed_ascwds_filtering_rule_column_values = [
```
Comment (MarionJHolloway) [inline projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_01_merge_metadata.py:135]: Can you add a comment here as to why this is being removed?

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/tests/fargate/utils/test_merge_utils.py [hunk 6]
```diff
@@ -64,3 +67,42 @@ def test_function_returns_expected_values(self, case):
...
+            lookback_fy_years=lookback_fy_years,
+            quarter_months=quarter_months,
+            col=col,
+        )
+
+        # Create test data around boundary
+        df = pl.DataFrame(
+            {
+                col: [
+                    date(2022, 3, 31),  # before monthly_start and quarterly rule does not match -> excluded
```
Comment (MarionJHolloway) [inline projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/tests/fargate/utils/test_merge_utils.py]: I think it would help to sort this column by date and keep the terminology you use to describe the test data consistant (e.g. before monthly start and old FY are the same thing I think?)

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/tests/fargate/utils/test_merge_utils.py [hunk 7]
```diff
@@ -64,3 +67,42 @@ def test_function_returns_expected_values(self, case):
         # Sanity check: correct expansion
         expected_rows = len(case.estimates_data) * 2  # mocked roles
         assert result_lf.collect().height == expected_rows
+
+
+class TestReducedDataFilterExpr(unittest.TestCase):
+    def test_expr_function_filters_rows_correctly(self):
+        # Given
```
Comment (MarionJHolloway) [inline projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/tests/fargate/utils/test_merge_utils.py]: These are all your default values, so I'd suggest parameterising these & your test/ expected data and then adding in a case with non-default options and a case for each arg passed as a null (separately).

**Review by MarionJHolloway — CHANGES_REQUESTED** [review]
A few small changes and some extra tests please :)

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/utils/merge_utils.py (hunk 3 above)
Comment (AchyutRatkanthiwar) [reply projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/utils/merge_utils.py:96]: Done

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/utils/merge_utils.py (hunk 4 above)
Comment (AchyutRatkanthiwar) [reply projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/utils/merge_utils.py]: Sorry, what do you mean by this?

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_01_merge_metadata.py (hunk 5 above)
Comment (AchyutRatkanthiwar) [reply projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/validate_01_merge_metadata.py:135]: yea, I mentioned it in PR description but I agree it will be better if it were added here as well. Added now

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/tests/fargate/utils/test_merge_utils.py (hunk 6 above)
Comment (AchyutRatkanthiwar) [reply projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/tests/fargate/utils/test_merge_utils.py]: Done

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/tests/fargate/utils/test_merge_utils.py (hunk 7 above)
Comment (AchyutRatkanthiwar) [reply projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/tests/fargate/utils/test_merge_utils.py]: Added in tests for them. Didnt add separate Null test case for `fy_start_month`,  `lookback_fy_years` and  `quarter_months` as they are not Nullable in function definition

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/utils/merge_utils.py (hunk 2 above)
Comment (AchyutRatkanthiwar) [reply projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/utils/merge_utils.py]: Done

File: projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/utils/merge_utils.py (hunk 4 above)
Comment (MarionJHolloway) [reply projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/utils/merge_utils.py]: today:date = today or date.today()

Comment (AchyutRatkanthiwar) [reply projects/_03_independent_cqc/_07_estimate_filled_posts_by_job_role/fargate/utils/merge_utils.py]: Done

**Review by MarionJHolloway — APPROVED** [review]
